### PR TITLE
Fix top-level speculation

### LINKED ISSUE #60248: Unnecessary cast diagnostic sometimes missing in top level programs
**Version Used**:  17.1.1 (also repros in main)

**Steps to Reproduce**:

1.  Create a new .net 6 console app
2.  Replace the contents of program.cs with 
```
int x = 1;
int y = (int)x;
int z = (int)1;

class C
{
    void M()
    {
        int i = 1;
        int j = (int)j;
        int k = (int)1;
    }
}
```
3.  Note the unnecessary cast diagnostic only on the cast in `z` but not on `y`.  Note that both unnecessary casts in method `M()` are marked as unnecessary (`j` and `k`).

**Expected Behavior**:
An unnecessary cast diagnostic on `y` as well.

### PR DESCRIPTION
**Pull request by astroC86**
Fixes #60248

Speculating in a global statement position used to fail because we had a null parent for the `StatementSyntax` (there isn't a parent `GlobalStatementSyntax` nor a `CompilationUnitSyntax` grand-parent nodes)

TODO

- [x] Will write compiler tests. **Done**

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=60274) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=60274) before we can accept your contribution.<br/><br/>:x: astroC86 [sign now](https://cla.dotnetfoundation.org/dotnet/roslyn?pullRequest=60274)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/roslyn?pullRequest=60274) it.</sub>

File: src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs [hunk 1]
```diff
@@ -9568,5 +9568,38 @@ void F<T>(T t)
...
+            var root = tree.GetRoot();
+            var model = compilation.GetSemanticModel(tree);
+            var nodeToSpeculate = SyntaxFactory.ParseStatement("int y = x;");
+
+            // Speculate inside a valid top-level position.
+            model.TryGetSpeculativeSemanticModel(root.DescendantNodes().Single(n => n is ExpressionStatementSyntax { Parent: GlobalStatementSyntax }).Span.End, nodeToSpeculate, out var speculativeModelInTopLevel);
+            var conversionInTopLevel = speculativeModelInTopLevel.GetConversion(nodeToSpeculate.DescendantTokens().Single(n => n.ValueText == "x").Parent);
+            Assert.Equal(ConversionKind.Identity, conversionInTopLevel.Kind);
+
+            // Speculate inside a valid top-level position.
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TopLevelStatementsTests.cs:9599]:
```suggestion
            // Speculate outside a top-level position.
```

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 5)

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review.

**Review by 333fred — APPROVED** [review]

**Comment by 333fred** [comment]
@astroC86 it looks like you'll need to merge `main` in and fix the conflict.

File: src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs [hunk 2]
```diff
@@ -12769,6 +12769,31 @@ static Delegate M(object o)
             }.RunAsync();
         }
 
+
+        [WorkItem(60248, "https://github.com/dotnet/roslyn/issues/60248")]
```
Comment (Youssef1313) [inline src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs:12773]:
```suggestion
        [WorkItem(60248, "https://github.com/dotnet/roslyn/issues/60248")
```

File: src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs [hunk 3]
```diff
@@ -12769,6 +12769,31 @@ static Delegate M(object o)
             }.RunAsync();
         }
 
+
```
Comment (Youssef1313) [inline src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs:12772]:
```suggestion
```

File: src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs [hunk 4]
```diff
@@ -12769,6 +12769,31 @@ static Delegate M(object o)
...
+",
+                LanguageVersion = LanguageVersion.CSharp10,
+                TestState =
+                {
+                    OutputKind = OutputKind.ConsoleApplication,
+                },
+            };
+
+            await test.RunAsync();
+          }
```
Comment (Youssef1313) [inline src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs:12795]: Is indentation bad here? I'm looking from mobile so not sure

Comment (333fred) [reply src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs:12794]: Yes, the indentation is bad.

**Review by 333fred — COMMENTED** [review]
Formatting needs to be fixed.

File: src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs [hunk 5]
```diff
@@ -12769,6 +12769,30 @@ static Delegate M(object o)
...
```
Comment (Youssef1313) [inline src/Analyzers/CSharp/Tests/RemoveUnnecessaryCast/RemoveUnnecessaryCastTests.cs:12794]:
```suggestion
        }
```

**Comment by Youssef1313** [comment]
@333fred @AlekseyTs Can you please restart integration tests here?

**Comment by 333fred** [comment]
Integration failures are known, hopefully to be addressed shortly. I'm going to force merge this.

**Comment by 333fred** [comment]
Thanks @astroC86!

**Pull request merged**